recipe-client-addon/bootstrap.js

```diff
@@ -29,69 +38,146 @@ const DEFAULT_PREFS = {
   "app.shield.optoutstudies.enabled": AppConstants.MOZ_DATA_REPORTING,
 };
 
-this.install = function() {};
-
-this.startup = function() {
-  // Initialize preference defaults before anything else happens.
-  const prefBranch = Services.prefs.getDefaultBranch("");
-  for (const [name, value] of Object.entries(DEFAULT_PREFS)) {
-    switch (typeof value) {
-      case "string":
-        prefBranch.setCharPref(name, value);
-        break;
-      case "number":
-        prefBranch.setIntPref(name, value);
-        break;
-      case "boolean":
-        prefBranch.setBoolPref(name, value);
-        break;
-      default:
-        throw new Error(`Invalid default preference type ${typeof value}`);
+// Logging
+const log = Log.repository.getLogger(BOOTSTRAP_LOGGER_NAME);
+log.addAppender(new Log.ConsoleAppender(new Log.BasicFormatter()));
+log.level = Services.prefs.getIntPref(PREF_LOGGING_LEVEL, Log.Level.Warn);
+
+this.Bootstrap = {
+  initShieldPrefs(defaultPrefs) {
+    const prefBranch = Services.prefs.getDefaultBranch("");
+    for (const [name, value] of Object.entries(defaultPrefs)) {
+      switch (typeof value) {
+        case "string":
+          prefBranch.setCharPref(name, value);
+          break;
+        case "number":
+          prefBranch.setIntPref(name, value);
+          break;
+        case "boolean":
+          prefBranch.setBoolPref(name, value);
+          break;
+        default:
+          throw new Error(`Invalid default preference type ${typeof value}`);
+      }
     }
-  }
+  },
 
-  ShieldRecipeClient.startup();
-};
+  initExperimentPrefs() {
+    const defaultBranch = Services.prefs.getDefaultBranch("");
+    const experimentBranch = Services.prefs.getBranch(STARTUP_EXPERIMENT_PREFS_BRANCH);
+
+    for (const childPrefName of experimentBranch.getChildList("")) {
+      const realPrefName = childPrefName.slice(1); // Remove leading dot
+      const realPrefType = defaultBranch.getPrefType(childPrefName);
+      const experimentPrefType = experimentBranch.getPrefType(childPrefName);
+
+      // TODO: This never passes?
```

Comment:
For some reason, as far as I can tell, `realPrefType` is _always_ `0` (`PREF_INVALID`) here. I marked the test as skipped just to get the suite to pass, but even if the test sets the pref value and outputs the right type before calling this function, the function still thinks the real pref doesn't exist and doesn't detect the value conflict.

Any insight on why `realPrefType` has the wrong value would be appreciated.

Reply:
`realPrefType` comes from querying the default branch. That wouldn't ever have any values for any of our prefs, right? That would explain why this is always `PREF_INVALID`.

Reply:
This is for experimental preferences, not Shield preferences. Experiment preferences would totally have default values.

Reply:
Ah, ok. I'm not sure it is safe to say that ll experiment preferences will have default branches, since some of the experiments are themselves add-ons. But this doesn't explain the `PREF_INVALID` thing.

Reply:
You're passing `childPrefName` instead of `realPrefName`, so instead of asking for, say, "browser.chromeURL", you're asking for ".browser.chromeURL", which doesn't exist.